File: Week_02/G20200343040243/group-anagrams.cpp
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;  
        // res vector的下标值
        int sub = 0;
        string tmp;
        // 判断排序后单词是否存在，即字母组成是否一致
        // string是排序后单词，int是在res vector中的下标(排序后不同的单词在res vector中下标值是递增的)
        unordered_map<string,int> work;

        for(auto str:strs) {
            tmp=str;
            sort(tmp.begin(),tmp.end());

            if(work.count(tmp)) {
               res[work[tmp]].push_back(str);
            } else {
                vector<string> vec(1,str);
                // 将未排序的字符串push到res vector末尾
                res.push_back(vec);
                // 记录排序完的字符串以及它在res vector中的位置，因为是res.push_back，所以在res vector的位置就是上一次push_back后面一个位置
                work[tmp] = sub++;
            }
        }
        return res;
    }
};

int main() {

    vector<string> arr = {"eat", "tea", "tan", "ate", "nat", "bat"};


    Solution s = Solution();
    vector<vector<string>> res = s.groupAnagrams(arr);

    for(vector<string> v : res) {
        for(string s : v) {
            cout << s << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Week_02/G20200343040115/GroupAnagrams.java
//给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。 
//
// 示例: 
//
// 输入: ["eat", "tea", "tan", "ate", "nat", "bat"],
//输出:
//[
//  ["ate","eat","tea"],
//  ["nat","tan"],
//  ["bat"]
//] 
//
// 说明： 
//
// 
// 所有输入均为小写字母。 
// 不考虑答案输出的顺序。 
// 
// Related Topics 哈希表 字符串


package com.Simba.algorithm007class01.leetcode.editor.cn.leetcode.editor.cn;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class GroupAnagrams {
    public static void main(String[] args) {
        Solution solution = new GroupAnagrams().new Solution();
        List<List<String>> list = new ArrayList<List<String>>();

        list = solution.groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"});
        for (List ll : list) {
            System.out.println(ll);

        }
    }


    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public List<List<String>> groupAnagrams(String[] strs) {
            //map将keystr存入当作key，与keystr相同的当作value
            HashMap<String,List<String>> map=new HashMap<>();
            for(String str:strs){
                char[]array=str.toCharArray();
                Arrays.sort(array);
                String keystr=String.valueOf(array);
                if(!map.containsKey(keystr)) map.put(keystr,new ArrayList<String>());
                map.get(keystr).add(str);
            }
            return new ArrayList<>(map.values());
        }
    }
//leetcode submit region end(Prohibit modification and deletion)
}